src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -583,6 +576,252 @@ private bool TryRemoveInvocation(string invocationId, out InvocationRequest irq)
+                                "The server disconnected before sending a handshake response");
+                        }
+                    }
+                    finally
+                    {
+                        _connection.Transport.Input.AdvanceTo(consumed, examined);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                // shutdown if we're unable to read handshake
+                Log.ErrorReceivingHandshakeResponse(_logger, ex);
+                throw;
+            }
+
+            Log.HandshakeComplete(_logger);
+        }
+
+        private async Task ReceiveLoop()
+        {
+            Log.ReceiveLoopStarting(_logger);
+
+            // Check if we need keep-alive
+            Timer timeoutTimer = null;
+            if (_connection.Features.Get<IConnectionInherentKeepAliveFeature>() == null)
+            {
+                Log.StartingServerTimeoutTimer(_logger, ServerTimeout);
+                timeoutTimer = new Timer(state =>
+                {
+                    if (!Debugger.IsAttached)
+                    {
+                        ((IConnection)state).Transport.Input.CancelPendingRead();
+                    }
+                }, _connection, ServerTimeout, Timeout.InfiniteTimeSpan);
+            }
+            else
+            {
+                Log.NotUsingServerTimeout(_logger);
+            }
+
+            Exception closeException = null;
+
+            try
+            {
+                while (true)
+                {
+                    var result = await _connection.Transport.Input.ReadAsync();
+                    var buffer = result.Buffer;
+                    var consumed = buffer.End; // TODO: Support partial messages
+                    var examined = buffer.End;
+
+                    try
+                    {
+                        if (buffer.IsEmpty && result.IsCompleted)
+                        {
+                            // The server disconnected
+                            break;
+                        }
+                        else if (result.IsCanceled)
+                        {
+                            // We aborted because the server timed-out
+                            closeException = new TimeoutException(
+                                $"Server timeout ({ServerTimeout.TotalMilliseconds:0.00}ms) elapsed without receiving a message from the server.");
+                            break;
+                        }
+
+                        if (timeoutTimer != null)
+                        {
+                            Log.ResettingKeepAliveTimer(_logger);
+                            timeoutTimer.Change(ServerTimeout, Timeout.InfiniteTimeSpan);
+                        }
+
+                        // We have data, process it
+                        var (close, exception) = await ProcessMessagesAsync(buffer);
+                        if (close)
```

Comment:
I think we’ll want to reduce the amount of async state machines here eventually.

Reply:
Agreed, but I'd rather keep the logic in easier-to-understand methods right now and tidy that up afterwards.

Reply:
Yep thats why I said eventually 😄